### PR TITLE
ENH: add option `on invalid="fix"` to `from wkb` and `from wkt`

### LINKED ISSUE #1342: ENH: expose fix_structure keyword in WKT reader (close linestrings automatically)
## Details of feature

With a polygon like so: `"POLYGON ((0 0, 1 0, 1 1, 0 1))"` (Not closed, missing the last `0 0` point)

Your geometry class will load the above shape just fine

```python
from shapely.geometry import Polygon
poly = Polygon([(0,0), (1,0), (1,1), (0,1)])
print(poly.wkt)
'POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0))' # adds the 0 0 to close the linestring automatically
```

But when loading the wkt, you get: `IllegalArgumentException: Points of LinearRing do not form a closed linestring`.

```python
from shapely import wkt
wkt.loads("POLYGON ((0 0, 1 0, 1 1, 0 1))")
# Throws with the above message.
```

## Proposal

Add a `strict=True` val (or something) to `wkt.loads`, that either just repairs this automatically, or at least loads the coords as-is and lets the user go in and make it valid.

```python
from shapely import wkt
poly = wkt.loads("POLYGON ((0 0, 1 0, 1 1, 0 1))", strict=False) # override strict to False, True by default
print(poly.wkt)
# If you repair automatically
'POLYGON ((0 0, 0 1, 1 1, 1 0, 0 0))'
# If you load as-is, letting the user decide how to repair it
'POLYGON ((0 0, 0 1, 1 1, 1 0))'
```

It's possible to have nested GeometryCollections, and if a polygon deep inside your wkt has this error, you're forced to use other WKT libraries to recurse down and repair it before shapely will be able to load it. This means you can't use any of the shapely library to separate out the WKT into individual shapes and fix the problem.

### PR DESCRIPTION
**Pull request by theroggy**
Add the option to "fix" invalid input geometries for `from_wkt` and `from_wkb`.

closes #1342

References:
- #1570
- https://github.com/geopandas/pyogrio/issues/423
- https://github.com/geopandas/geopandas/issues/2791
- https://github.com/geopandas/geopandas/discussions/3336

### CODE REVIEW

File: src/ufuncs.c [hunk 1]
```diff
@@ -3066,6 +3066,16 @@ static void from_wkb_func(char** args, const npy_intp* dimensions, const npy_int
     goto finish;
   }
 
+  /* If on_invalid is 3, try to fix invalid geometries */
+  if (on_invalid == 3) {
+    #if GEOS_SINCE_3_11_0
+      GEOSWKBReader_setFixStructure_r(ctx, reader, 1);
+    #else
+      PyErr_Format(PyExc_ValueError, "on_invalid='fix' only supported for GEOS >= 3.11");
```
Comment (brendan-ward) [inline src/ufuncs.c:3074]: This should probably have a test case too, skipped for GEOS >= 3.11

File: shapely/io.py [hunk 2]
```diff
@@ -269,11 +269,21 @@ def from_wkt(geometry, on_invalid="raise", **kwargs):
...
+        points for the geometry type. For a thorough check, use
+        :func:`is_valid` after conversion to geometries. Valid options are:
+
+        - raise: an exception will be raised if any input geometry is invalid.
         - warn: a warning will be raised and invalid WKT geometries will be
           returned as ``None``.
-        - ignore: invalid WKT geometries will be returned as ``None`` without a warning.
+        - ignore: invalid geometries will be returned as ``None`` without a
+          warning.
+        - fix: an effort is made to fix invalid input geometries. If this is not
```
Comment (brendan-ward) [inline shapely/io.py:283]:
```suggestion
        - fix: an effort is made to fix invalid input geometries (currently just unclosed rings). If this is not
```

(per the GEOS docs)

File: src/ufuncs.c (hunk 1 above)
Comment (theroggy) [reply src/ufuncs.c:3074]: Added some tests for this case...

**Review by brendan-ward — APPROVED** [review]
Looks good to me, thanks!

**Review by mwtoews — APPROVED** [review]
Thanks, looks good! Xref to upstream implementation here: https://github.com/libgeos/geos/pull/639
with original discussion here: https://github.com/libgeos/geos/issues/587

Also note that this PR fixes #1342

**Comment by jorisvandenbossche** [comment]
Didn't think about it before, but rereading the docstring for the new option: "If this is not possible, they are returned as ``None`` without a warning.". 
We _could_ also opt for still raising an error or issuing a warning in that case. Of course we don't want to add all possible combinations as options for this keyword, but I do wonder what generally would be the most desirable behaviour if you want things to get fixed automatically (because now with silent ignoring, you're still not sure that the invalid geometries were fixed without checking for Nones)

**Comment by theroggy** [comment]
It is kind of complicated. There are two main cases:
- the geometry could not be fixed: None result, and data is lost. For a WKT in this situation we could still log the coordinates in a warning, so the user can still salvage something manually, for a WKB this is not as easy...
- the geometry was fixed: in many (most) cases this will lead to an "invalid geometry". If `make_valid` is applied in many cases the result will be of a lower dimensionality than the original geometry, so a polygon becomes a line, a line becomes a point.

So, even if the geometry was fixed, it will still often need some manual further treatment to get to a usable result...

I also did some extra tests, and e.g. a multipolygon with one valid ring, and one ring with only one point will not be fixed, so will be entirely lost... (=None), so there is still quite some significant data loss possible.

All this doesn't mean that raising warnings is not useful... maybe rather that when warnings are given, they are probably just as useful for the fixed geometries...

**Comment by theroggy** [comment]
Not sure what you mean with fancy... but indeed, it is a native GEOS feature from the C API being exposed, so I don't think there are alternatives to shapely to expose this functionality.

It is also just an extra option for an existing parameter, so the shapely API didn't become that much more complicated to support this.

**Comment by sgillies** [comment]
@theroggy my quibble is really with the GEOS C API design. I'm sorry about aiming my comment at you.

**Pull request merged**